File: nodes/custom/polygon/polygon_crypto_universe_node.py
import logging
from datetime import datetime
from typing import Any

import httpx
import pytz

from core.api_key_vault import APIKeyVault
from core.types_registry import AssetClass, AssetSymbol, get_type
from nodes.base.base_node import Base
from services.polygon_service import (
    massive_build_snapshot_tickers,
    massive_fetch_snapshot,
    massive_get_numeric_from_dict,
    massive_parse_ticker_for_market,
)
from services.time_utils import utc_timestamp_flex_to_et_datetime

logger = logging.getLogger(__name__)


class PolygonCryptoUniverse(Base):
    """
    A node that fetches crypto symbols from the Massive.com API (formerly Polygon.io) and filters them
    based on the provided parameters.

    Note: Polygon.io has rebranded to Massive.com. The API endpoints have been updated
    to use api.massive.com, but the API routes remain unchanged.

    Endpoint: https://api.massive.com/v2/snapshot/locale/global/markets/crypto/tickers

    Crypto markets operate 24/7, always using current intraday data from the 'day' bar.
    """

    inputs = {"filter_symbols": get_type("AssetSymbolList") | None}
    outputs = {"symbols": get_type("AssetSymbolList")}
    required_keys = ["POLYGON_API_KEY"]
    # Common stablecoin tickers to filter out
    STABLECOIN_TICKERS = {
        "USDT",  # Tether
        "USDC",  # USD Coin
        "DAI",   # Dai
        "BUSD",  # Binance USD
        "TUSD",  # TrueUSD
        "USDP",  # Pax Dollar
        "USDD",  # USDD
        "GUSD",  # Gemini Dollar
        "HUSD",  # HUSD
        "USDX",  # USDX
        "FRAX",  # Frax
        "LUSD",  # Liquity USD
        "SUSD",  # sUSD
        "USDS",  # USDS
        "USN",   # USN
        "USDK",  # USDK
        "EURS",  # STASIS EURS
        "EURT",  # Tether EUR
        "GBPT",  # Tether GBP
    }

    params_meta = [
        {
            "name": "exclude_stablecoins",
            "type": "boolean",
            "default": True,
            "label": "Exclude Stablecoins",
            "description": "Exclude stablecoins (USDT, USDC, DAI, etc.) from results",
        },
        {
            "name": "min_change_perc",
            "type": "number",
            "default": None,
            "label": "Min Change",
            "unit": "%",
            "description": "Minimum daily percentage change (e.g., 5 for 5%)",
            "step": 0.01,
        },
        {
            "name": "max_change_perc",
            "type": "number",
            "default": None,
            "label": "Max Change",
            "unit": "%",
            "description": "Maximum daily percentage change (e.g., 10 for 10%)",
            "step": 0.01,
        },
        {
            "name": "min_volume",
            "type": "number",
            "default": None,
            "label": "Min Volume",
            "unit": "shares/contracts",
            "description": "Minimum daily trading volume",
        },
        {
            "name": "min_price",
            "type": "number",
            "default": None,
            "label": "Min Price",
            "unit": "USD",
            "description": "Minimum closing price in USD",
        },
        {
            "name": "max_price",
            "type": "number",
            "default": 1000000,
            "label": "Max Price",
            "unit": "USD",
            "description": "Maximum closing price in USD",
        },
        {
            "name": "max_snapshot_delay_minutes",
            "type": "combo",
            "default": "5min",
            "label": "Max Snapshot Delay",
            "description": "Maximum allowed delay in snapshot 'updated' timestamp; None = no filter",
            "options": ["None (no filter)", "5min", "15min", "120min"],
        },
    ]
    default_params = {
        "exclude_stablecoins": True,  # Default to excluding stablecoins
    }

    async def _execute_impl(self, inputs: dict[str, Any]) -> dict[str, Any]:
        try:
            filter_symbols = self.collect_multi_input("filter_symbols", inputs)
            symbols = await self._fetch_symbols(filter_symbols)
            return {"symbols": symbols}
        except Exception as e:
            logger.error(f"PolygonCryptoUniverse node {self.id} failed: {str(e)}", exc_info=True)
            raise

    async def _fetch_symbols(
        self, filter_symbols: list[AssetSymbol] | None = None
    ) -> list[AssetSymbol]:
        api_key = APIKeyVault().get("POLYGON_API_KEY")
        if not api_key:
            raise ValueError("POLYGON_API_KEY is required but not set in vault")

        market = "crypto"
        locale, markets = "global", "crypto"

        filter_ticker_strings: list[str] | None = None
        if filter_symbols:
            filter_ticker_strings = massive_build_snapshot_tickers(filter_symbols)

        # Use longer timeout for snapshot fetch (can be slow with many tickers)
        timeout = httpx.Timeout(30.0, connect=10.0)
        async with httpx.AsyncClient(timeout=timeout) as client:
            tickers_data = await massive_fetch_snapshot(
                client,
                api_key,
                locale,
                markets,
                market,
                filter_ticker_strings,
                False,
            )

            filter_params = self._extract_filter_params()
            self._validate_filter_params(filter_params)

            symbols = self._process_tickers(tickers_data, market, filter_params)

        return symbols

    def _get_numeric_param(self, param_name: str) -> float | None:
        """Extract and validate numeric parameter."""
        param_raw = self.params.get(param_name)
        if isinstance(param_raw, int | float):
            return float(param_raw)
        return None

    def _extract_filter_params(self) -> dict[str, float | None]:
        """Extract all filter parameters."""
        # Parse max_snapshot_delay_minutes from combo string
        delay_raw = self.params.get("max_snapshot_delay_minutes")
        max_snapshot_delay_minutes: float | None = None
        if delay_raw == "None (no filter)":
            max_snapshot_delay_minutes = None
        elif isinstance(delay_raw, str):
            if delay_raw.endswith("min"):
                try:
                    max_snapshot_delay_minutes = float(delay_raw[:-3])
                except ValueError:
                    logger.warning(f"Invalid delay string '{delay_raw}'; no filter applied")
        # Fallback to old numeric if present (for compatibility)
        elif isinstance(delay_raw, (int, float)):
            max_snapshot_delay_minutes = float(delay_raw)

        return {
            "min_change_perc": self._get_numeric_param("min_change_perc"),
            "max_change_perc": self._get_numeric_param("max_change_perc"),
            "min_volume": self._get_numeric_param("min_volume"),
            "min_price": self._get_numeric_param("min_price"),
            "max_price": self._get_numeric_param("max_price"),
            "max_snapshot_delay_minutes": max_snapshot_delay_minutes,
        }

    def _validate_filter_params(self, filter_params: dict[str, float | None]) -> None:
        """Validate filter parameters for logical consistency."""
        min_change_perc = filter_params["min_change_perc"]
        max_change_perc = filter_params["max_change_perc"]
        if min_change_perc is not None and max_change_perc is not None:
            if min_change_perc > max_change_perc:
                raise ValueError("min_change_perc cannot be greater than max_change_perc")

    def _process_tickers(
        self,
        tickers_data: list[dict[str, Any]],
        market: str,
        filter_params: dict[str, float | None],
    ) -> list[AssetSymbol]:
        """Process ticker data and apply filters to produce AssetSymbol list."""
        symbols: list[AssetSymbol] = []
        total_tickers = len(tickers_data)
        filtered_stale = 0

        current_utc = datetime.now(pytz.UTC)

        for ticker_item in tickers_data:
            ticker = self._extract_ticker(ticker_item)
            if not ticker:
                continue

            # Filter: Only include USD-quoted crypto pairs
            if not self._is_usd_quoted(ticker, market):
                continue

            # Filter: Exclude stablecoins if enabled
            if self._should_exclude_stablecoin(ticker):
                continue

            ticker_data = self._extract_ticker_data(ticker_item, market)
            if not ticker_data:
                continue

            # Additional stablecoin filter: exclude if price is very close to $1.00
            # (catches stablecoins not in our list)
            if self._should_exclude_stablecoin_by_price(ticker_data):
                continue

            # New: Check snapshot update delay if filter enabled
            max_delay_min = filter_params.get("max_snapshot_delay_minutes")
            if max_delay_min is not None:
                updated_raw = ticker_item.get("updated")
                if isinstance(updated_raw, int):
                    updated_et = utc_timestamp_flex_to_et_datetime(updated_raw)
                    if updated_et is None:
                        # Invalid parse: treat as stale
                        filtered_stale += 1
                        logger.warning(
                            f"Filtered ticker {ticker} due to invalid 'updated' timestamp"
                        )
                        continue
                    updated_utc = updated_et.astimezone(pytz.UTC)
                    delay_minutes = (current_utc - updated_utc).total_seconds() / 60
                    if delay_minutes > max_delay_min:
                        filtered_stale += 1
                        logger.info(
                            f"Filtered stale snapshot for {ticker}: {delay_minutes:.1f} min delay"
                        )
                        continue
                else:
                    # No updated timestamp: treat as stale
                    filtered_stale += 1
                    logger.warning(f"Filtered ticker {ticker} due to missing 'updated' timestamp")
                    continue

            if not self._passes_filters(ticker_data, filter_params):
                continue

            symbol = self._create_asset_symbol(ticker, ticker_item, market, ticker_data)
            symbols.append(symbol)

        # Update progress with stale filter info if applied
        stale_msg = f"; filtered {filtered_stale} stale snapshots" if filtered_stale > 0 else ""
        self.report_progress(
            95.0, f"Completed: {len(symbols)} symbols from {total_tickers} tickers{stale_msg}"
        )
        return symbols

    def _extract_ticker(self, ticker_item: dict[str, Any]) -> str | None:
        """Extract ticker string from ticker item."""
        ticker_value = ticker_item.get("ticker")
        return ticker_value if isinstance(ticker_value, str) else None

    def _is_usd_quoted(self, ticker: str, market: str) -> bool:
        """Check if ticker is USD-quoted (for crypto filtering)."""
        if market != "crypto":
            return True  # No filtering for non-crypto markets

        _, quote_currency = massive_parse_ticker_for_market(ticker, market)
        # Only include USD-quoted pairs; None means parsing failed, exclude for safety
        return quote_currency == "USD"

    def _should_exclude_stablecoin(self, ticker: str) -> bool:
        """Check if ticker should be excluded as a stablecoin."""
        exclude_stablecoins = self.params.get("exclude_stablecoins", True)
        if not exclude_stablecoins:
            return False

        # Parse ticker to get base symbol
        base_ticker, _ = massive_parse_ticker_for_market(ticker, "crypto")
        if not base_ticker:
            return False

        # Check if base ticker is in stablecoin list (case-insensitive)
        return base_ticker.upper() in self.STABLECOIN_TICKERS

    def _should_exclude_stablecoin_by_price(self, ticker_data: dict[str, Any]) -> bool:
        """Check if ticker should be excluded based on price being very close to $1.00 (stablecoin indicator)."""
        exclude_stablecoins = self.params.get("exclude_stablecoins", True)
        if not exclude_stablecoins:
            return False

        price_raw = ticker_data.get("price")
        if not isinstance(price_raw, (int, float)):
            return False

        price = float(price_raw)
        # Exclude if price is between $0.99 and $1.01 (typical stablecoin range)
        # This catches stablecoins not in our explicit list
        return 0.99 <= price <= 1.01

    def _extract_ticker_data(
        self, ticker_item: dict[str, Any], market: str
    ) -> dict[str, Any] | None:
        """Extract price, volume, and change percentage from ticker data.

        For crypto, markets are 24/7, always using current intraday data from the 'day' bar.
        """
        day_value = ticker_item.get("day")
        if not isinstance(day_value, dict):
            ticker_log = ticker_item.get("ticker")
            if isinstance(ticker_log, str):
                logger.warning(f"Invalid 'day' data for ticker {ticker_log}")
            else:
                logger.warning("Invalid 'day' data for ticker (invalid ticker key)")
            day: dict[str, Any] = {}
        else:
            day = day_value

        # Extract todaysChangePerc (always available, reflects change since prevDay.c)
        change_perc_raw = ticker_item.get("todaysChangePerc")
        if isinstance(change_perc_raw, int | float):
            todays_change_perc = float(change_perc_raw)
        else:
            ticker_log = ticker_item.get("ticker")
            if isinstance(ticker_log, str):
                logger.warning(f"Missing or invalid todaysChangePerc for ticker {ticker_log}")
            else:
                logger.warning(
                    "Missing or invalid todaysChangePerc for ticker (invalid ticker key)"
                )
            todays_change_perc = None

        # Today (always)
        price = massive_get_numeric_from_dict(day, "c", 0.0)
        volume = massive_get_numeric_from_dict(day, "v", 0.0)
        change_perc = todays_change_perc  # API-provided, from prevDay to day.c
        data_source = "day"

        if price <= 0:
            ticker_log = ticker_item.get("ticker")
            if isinstance(ticker_log, str):
                logger.warning(f"Invalid price (<=0) for ticker {ticker_log}")
            else:
                logger.warning("Invalid price (<=0) for ticker (invalid ticker key)")
            return None

        return {
            "price": price,
            "volume": volume,
            "change_perc": change_perc,
            "data_source": data_source,  # For metadata
        }

    def _passes_filters(
        self, ticker_data: dict[str, Any], filter_params: dict[str, float | None]
    ) -> bool:
        """Check if ticker data passes all configured filters."""
        # Type guards for ticker_data values
        price_raw = ticker_data.get("price")
        if not isinstance(price_raw, int | float):
            return False
        price = float(price_raw)

        volume_raw = ticker_data.get("volume")
        if not isinstance(volume_raw, int | float):
            return False
        volume = float(volume_raw)

        change_perc_raw = ticker_data.get("change_perc")
        change_perc: float | None = None
        if isinstance(change_perc_raw, int | float):
            change_perc = float(change_perc_raw)

        min_change_perc = filter_params["min_change_perc"]
        max_change_perc = filter_params["max_change_perc"]
        min_volume = filter_params["min_volume"]
        min_price = filter_params["min_price"]
        max_price = filter_params["max_price"]

        if change_perc is not None:
            if min_change_perc is not None and change_perc < min_change_perc:
                return False
            if max_change_perc is not None and change_perc > max_change_perc:
                return False

        if min_volume is not None and volume < min_volume:
            return False

        if min_price is not None and price < min_price:
            return False

        if max_price is not None and price > max_price:
            return False

        return True

    def _create_asset_symbol(
        self,
        ticker: str,
        ticker_item: dict[str, Any],
        market: str,
        ticker_data: dict[str, Any],
    ) -> AssetSymbol:
        """Create AssetSymbol from ticker data."""
        base_ticker, quote_currency = massive_parse_ticker_for_market(ticker, market)

        asset_class = AssetClass.CRYPTO

        # Type guards for metadata
        data_source_raw = ticker_data.get("data_source")
        data_source = data_source_raw if isinstance(data_source_raw, str) else "unknown"

        change_perc_raw = ticker_data.get("change_perc")
        change_available = isinstance(change_perc_raw, int | float)

        metadata = {
            "original_ticker": ticker,
            "snapshot": ticker_item,
            "market": market,
            "data_source": data_source,
            "change_available": change_available,
        }

        return AssetSymbol(
            ticker=base_ticker,
            asset_class=asset_class,
            quote_currency=quote_currency,
            metadata=metadata,
        )

File: packages/core/src/nodes/custom/polygon/polygon-crypto-universe-node.ts
// src/nodes/custom/polygon/polygon-crypto-universe-node.ts
// Translated from: nodes/custom/polygon/polygon_crypto_universe_node.py

import { Base } from '../../base/base-node';
import { getVault } from '../../../core/api-key-vault';
import {
  AssetClass,
  AssetSymbol,
  InstrumentType,
  NodeCategory,
  NodeUIConfig,
  ParamMeta,
  getType,
} from '../../../core/types';
import {
  massiveBuildSnapshotTickers,
  massiveFetchSnapshot,
  massiveGetNumericFromDict,
  massiveParseTickerForMarket,
} from '../../../services/polygon-service';

interface FilterParams {
  min_change_perc: number | null;
  max_change_perc: number | null;
  min_volume: number | null;
  min_price: number | null;
  max_price: number | null;
  max_snapshot_delay_minutes: number | null;
}

interface TickerData {
  price: number;
  volume: number;
  change_perc: number | null;
  data_source: string;
}

/**
 * A node that fetches crypto symbols from the Massive.com API (formerly Polygon.io)
 * and filters them based on the provided parameters.
 *
 * Endpoint: https://api.massive.com/v2/snapshot/locale/global/markets/crypto/tickers
 *
 * Crypto markets operate 24/7, always using current intraday data from the 'day' bar.
 */
export class PolygonCryptoUniverse extends Base {
  static inputs = { filter_symbols: getType('AssetSymbolList') };
  static outputs = { symbols: getType('AssetSymbolList') };
  static required_keys = ['POLYGON_API_KEY'];
  static uiConfig: NodeUIConfig = {
    size: [280, 140],
    displayResults: false,
    resizable: true,
  };

  static paramsMeta: ParamMeta[] = [
    {
      name: 'min_change_perc',
      type: 'number',
      default: undefined,
      label: 'Min Change',
      unit: '%',
      description: 'Minimum daily percentage change (e.g., 5 for 5%)',
      step: 0.01,
    },
    {
      name: 'max_change_perc',
      type: 'number',
      default: undefined,
      label: 'Max Change',
      unit: '%',
      description: 'Maximum daily percentage change (e.g., 10 for 10%)',
      step: 0.01,
    },
    {
      name: 'min_volume',
      type: 'number',
      default: undefined,
      label: 'Min Volume',
      unit: 'shares/contracts',
      description: 'Minimum daily trading volume',
    },
    {
      name: 'min_price',
      type: 'number',
      default: undefined,
      label: 'Min Price',
      unit: 'USD',
      description: 'Minimum closing price in USD',
    },
    {
      name: 'max_price',
      type: 'number',
      default: 1000000,
      label: 'Max Price',
      unit: 'USD',
      description: 'Maximum closing price in USD',
    },
    {
      name: 'max_snapshot_delay_minutes',
      type: 'combo',
      default: '5min',
      label: 'Max Snapshot Delay',
      description:
        "Maximum allowed delay in snapshot 'updated' timestamp; None = no filter",
      options: ['None (no filter)', '5min', '15min', '120min'],
    },
  ];

  static defaultParams = {};

  static CATEGORY = NodeCategory.MARKET;

  protected async executeImpl(
    inputs: Record<string, unknown>
  ): Promise<Record<string, unknown>> {
    try {
      const filterSymbolsRaw = inputs.filter_symbols;
      const filterSymbols: AssetSymbol[] = Array.isArray(filterSymbolsRaw) ? filterSymbolsRaw : [];
      const symbols = await this.fetchSymbols(filterSymbols.length > 0 ? filterSymbols : null);
      return { symbols };
    } catch (error) {
      const errorMsg = error instanceof Error ? error.message : String(error);
      console.error(`PolygonCryptoUniverse node ${this.id} failed: ${errorMsg}`);
      throw error;
    }
  }

  private async fetchSymbols(
    filterSymbols: AssetSymbol[] | null
  ): Promise<AssetSymbol[]> {
    const apiKey = getVault().get('POLYGON_API_KEY');
    if (!apiKey) {
      throw new Error('POLYGON_API_KEY is required but not set in vault');
    }

    const market = 'crypto';
    const locale = 'global';
    const markets = 'crypto';

    let filterTickerStrings: string[] | null = null;
    if (filterSymbols && filterSymbols.length > 0) {
      filterTickerStrings = massiveBuildSnapshotTickers(filterSymbols);
    }

    const tickersData = await massiveFetchSnapshot(
      apiKey,
      locale,
      markets,
      market,
      filterTickerStrings,
      false
    );

    const filterParams = this.extractFilterParams();
    this.validateFilterParams(filterParams);

    const symbols = this.processTickers(tickersData, market, filterParams);

    return symbols;
  }

  private getNumericParam(paramName: string): number | null {
    const paramRaw = this.params[paramName];
    return typeof paramRaw === 'number' ? paramRaw : null;
  }

  private extractFilterParams(): FilterParams {
    // Parse max_snapshot_delay_minutes from combo string
    const delayRaw = this.params.max_snapshot_delay_minutes;
    let maxSnapshotDelayMinutes: number | null = null;

    if (delayRaw === 'None (no filter)') {
      maxSnapshotDelayMinutes = null;
    } else if (typeof delayRaw === 'string' && delayRaw.endsWith('min')) {
      const parsed = parseFloat(delayRaw.slice(0, -3));
      if (!isNaN(parsed)) {
        maxSnapshotDelayMinutes = parsed;
      }
    } else if (typeof delayRaw === 'number') {
      maxSnapshotDelayMinutes = delayRaw;
    }

    return {
      min_change_perc: this.getNumericParam('min_change_perc'),
      max_change_perc: this.getNumericParam('max_change_perc'),
      min_volume: this.getNumericParam('min_volume'),
      min_price: this.getNumericParam('min_price'),
      max_price: this.getNumericParam('max_price'),
      max_snapshot_delay_minutes: maxSnapshotDelayMinutes,
    };
  }

  private validateFilterParams(filterParams: FilterParams): void {
    const { min_change_perc, max_change_perc } = filterParams;
    if (
      min_change_perc !== null &&
      max_change_perc !== null &&
      min_change_perc > max_change_perc
    ) {
      throw new Error('min_change_perc cannot be greater than max_change_perc');
    }
  }

  private processTickers(
    tickersData: Record<string, unknown>[],
    market: string,
    filterParams: FilterParams
  ): AssetSymbol[] {
    const symbols: AssetSymbol[] = [];
    const totalTickers = tickersData.length;
    let filteredStale = 0;

    const currentUtc = Date.now();

    for (const tickerItem of tickersData) {
      const ticker = this.extractTicker(tickerItem);
      if (!ticker) continue;

      // Filter: Only include USD-quoted crypto pairs
      if (!this.isUsdQuoted(ticker, market)) {
        continue;
      }

      const tickerData = this.extractTickerData(tickerItem, market);
      if (!tickerData) continue;

      // Check snapshot update delay if filter enabled
      const maxDelayMin = filterParams.max_snapshot_delay_minutes;
      if (maxDelayMin !== null) {
        const updatedRaw = tickerItem.updated;
        if (typeof updatedRaw === 'number') {
          const updatedUtc = this.parseTimestampFlex(updatedRaw);
          if (updatedUtc === null) {
            filteredStale++;
            console.warn(`Filtered ticker ${ticker} due to invalid 'updated' timestamp`);
            continue;
          }
          const delayMinutes = (currentUtc - updatedUtc) / 60000;
          if (delayMinutes > maxDelayMin) {
            filteredStale++;
            console.info(
              `Filtered stale snapshot for ${ticker}: ${delayMinutes.toFixed(1)} min delay`
            );
            continue;
          }
        } else {
          filteredStale++;
          console.warn(`Filtered ticker ${ticker} due to missing 'updated' timestamp`);
          continue;
        }
      }

      if (!this.passesFilters(tickerData, filterParams)) {
        continue;
      }

      const symbol = this.createAssetSymbol(ticker, tickerItem, market, tickerData);
      symbols.push(symbol);
    }

    // Update progress with stale filter info if applied
    const staleMsg = filteredStale > 0 ? `; filtered ${filteredStale} stale snapshots` : '';
    this.reportProgress(
      95.0,
      `Completed: ${symbols.length} symbols from ${totalTickers} tickers${staleMsg}`
    );

    return symbols;
  }

  private extractTicker(tickerItem: Record<string, unknown>): string | null {
    const tickerValue = tickerItem.ticker;
    return typeof tickerValue === 'string' ? tickerValue : null;
  }

  private isUsdQuoted(ticker: string, market: string): boolean {
    if (market !== 'crypto') {
      return true;
    }

    const [, quoteCurrency] = massiveParseTickerForMarket(ticker, market);
    return quoteCurrency === 'USD';
  }

  private extractTickerData(
    tickerItem: Record<string, unknown>,
    _market: string
  ): TickerData | null {
    const dayValue = tickerItem.day;
    const day: Record<string, unknown> =
      typeof dayValue === 'object' && dayValue !== null
        ? (dayValue as Record<string, unknown>)
        : {};

    // Extract todaysChangePerc
    const changePercRaw = tickerItem.todaysChangePerc;
    const todaysChangePerc =
      typeof changePercRaw === 'number' ? changePercRaw : null;

    // Crypto uses current day data (24/7 market)
    const price = massiveGetNumericFromDict(day, 'c', 0.0);
    const volume = massiveGetNumericFromDict(day, 'v', 0.0);
    const changePerc = todaysChangePerc;
    const dataSource = 'day';

    if (price <= 0) {
      const tickerLog = tickerItem.ticker;
      console.warn(
        `Invalid price (<=0) for ticker ${typeof tickerLog === 'string' ? tickerLog : '(unknown)'}`
      );
      return null;
    }

    return {
      price,
      volume,
      change_perc: changePerc,
      data_source: dataSource,
    };
  }

  private passesFilters(tickerData: TickerData, filterParams: FilterParams): boolean {
    const { price, volume, change_perc } = tickerData;
    const { min_change_perc, max_change_perc, min_volume, min_price, max_price } =
      filterParams;

    if (change_perc !== null) {
      if (min_change_perc !== null && change_perc < min_change_perc) {
        return false;
      }
      if (max_change_perc !== null && change_perc > max_change_perc) {
        return false;
      }
    }

    if (min_volume !== null && volume < min_volume) {
      return false;
    }

    if (min_price !== null && price < min_price) {
      return false;
    }

    if (max_price !== null && price > max_price) {
      return false;
    }

    return true;
  }

  private createAssetSymbol(
    ticker: string,
    tickerItem: Record<string, unknown>,
    market: string,
    tickerData: TickerData
  ): AssetSymbol {
    const [baseTicker, quoteCurrency] = massiveParseTickerForMarket(ticker, market);
    const assetClass = AssetClass.CRYPTO;

    const dataSource = tickerData.data_source;
    const changeAvailable = tickerData.change_perc !== null;

    const metadata = {
      original_ticker: ticker,
      snapshot: tickerItem,
      market,
      data_source: dataSource,
      change_available: changeAvailable,
    };

    return new AssetSymbol(
      baseTicker,
      assetClass,
      quoteCurrency ?? undefined,
      InstrumentType.SPOT,
      metadata
    );
  }

  /**
   * Parse timestamp with auto-detection of unit (s, ms, ns).
   */
  private parseTimestampFlex(ts: number): number | null {
    if (ts <= 0) {
      return null;
    }

    const tsStr = String(ts);
    const digitCount = tsStr.length;

    // Try to determine unit based on digit count
    let divisor: number;
    if (digitCount < 11) {
      divisor = 1000; // seconds -> ms
    } else if (digitCount >= 11 && digitCount <= 14) {
      divisor = 1; // already ms
    } else if (digitCount >= 15 && digitCount <= 19) {
      divisor = 1000000; // ns -> ms
    } else {
      divisor = 1; // default to ms
    }

    const timestampMs = ts / divisor;
    const date = new Date(timestampMs);

    // Validate: Unix epoch range (1970-2100)
    const year = date.getUTCFullYear();
    if (year >= 1970 && year <= 2100) {
      return timestampMs;
    }

    // Try other units
    const candidates = [
      { unit: 'ns', divisor: 1000000 },
      { unit: 'ms', divisor: 1 },
      { unit: 's', divisor: 0.001 },
    ];

    for (const { divisor: d } of candidates) {
      const tryMs = ts / d;
      const tryDate = new Date(tryMs);
      const tryYear = tryDate.getUTCFullYear();
      if (tryYear >= 1970 && tryYear <= 2100) {
        return tryMs;
      }
    }

    return null;
  }
}
